torch/nn/_functions/thnn/normalization.py

```diff
@@ -135,4 +135,101 @@ def backward(self, grad_output):
         return grad_input
 
 
+class SyncBatchNorm(Function):
+
+    @staticmethod
+    def forward(self, input, weight, bias, running_mean, running_var, eps, momentum, process_group=None):
+        input = input.contiguous()
+
+        # calcualte mean/invstd for input.
+        mean, invstd = torch.batch_norm_stats(input, eps)
+        if torch.distributed.is_initialized():
+            world_size = 0
+            if process_group:
+                world_size = torch.distributed.get_world_size(process_group)
+            else:
+                process_group = torch.distributed.get_default_group()
+                world_size = torch.distributed.get_world_size()
+            mean_all = torch.empty(world_size, mean.size(0), dtype=mean.dtype, device=mean.device)
+            invstd_all = torch.empty(world_size, invstd.size(0), dtype=invstd.dtype, device=invstd.device)
+            mean_l = [mean_all.narrow(0, i, 1) for i in range(world_size)]
+            invstd_l = [invstd_all.narrow(0, i, 1) for i in range(world_size)]
+            # using all_gather instead of all reduce so we can calculate mean/var in one go
+            torch.distributed.all_gather(mean_l, mean, process_group)
```

Comment:
less nccl calls so technically yes. Realistically speaking the actual communication part is tiny in the overall timeline. The synchronization might have some impact here, but since we have two consecutive all_gather/all_reduce, combining the two calls doesn't really help.